Add Matter climate running state heat fan and cool fan

virtualbitzz opened the pull request:
Add support for ThermostatRunningState values 5 for Heat+FAN and 6 Cool+FAN

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
Adds support for new values for ThermostatRunningState (which map to hvac_actions). This change recognizes new values 5 for HEAT+FAN and 6 for COOL+FAN and maps them to Heat and Cool respectively. 
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

virtualbitzz committed:
Update climate.py

Add support for ThermostatRunningState values 5 for Heat+FAN and 6 Cool+FAN

epenet commented:
cc @hidaris who added initial support via #95434

virtualbitzz commented on homeassistant/components/matter/climate.py:
```diff
@@ -298,11 +300,11 @@ def _update_from_device(self) -> None:
             ):
                 match running_state_value:
```
Unfortunately I've reached the limits of my programming capabilities. I have no idea how to implement your suggestions @epenet. I can close this and leave the feature request open. Please advise.

StephenR0 replied:
I would leave it open.  The developers will figure it out.

virtualbitzz committed:
Update climate.py

Replaced match running_state_value: logic

virtualbitzz committed:
Update climate.py

Removed combos for ThermostatRunningState

virtualbitzz reviewed (commented):
I _think_ I did what was asked.

virtualbitzz committed:
Update homeassistant/components/matter/climate.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

virtualbitzz committed:
Update test_climate.py

Add values for new ThermostatRunningState to >> hvac_actions mapping. Updated incorrect mapping

virtualbitzz committed:
Update test_climate.py

fixed typo on numerical value

virtualbitzz committed:
Update test_climate.py

Fixed indentation typo

epenet reviewed (approved):
LGTM 👍 

But since I helped write the code I think it's better to have a second opinion
Also, not sure if it should go in patch release

virtualbitzz commented on tests/components/matter/test_climate.py:
```diff
@@ -125,7 +137,7 @@ async def test_thermostat_base(
     state = hass.states.get("climate.longan_link_hvac")
     assert state
-    assert state.attributes["hvac_action"] == HVACAction.OFF
+    assert state.attributes["hvac_action"] == HVACAction.COOLING
```
I've noticed on 2025.8 that when values 5 or 6 are present that hvac_action maps to "OFF"

<img width="211" height="128" alt="image" src="https://github.com/user-attachments/assets/ad678747-42dc-441f-a7d9-5f190ec29dda" />

virtualbitzz replied:
@epenet I'm getting a KeyError when it's running its checks. It's complaining about the zero value. It doesn't appear to complain about 128.

virtualbitzz committed:
Update test_climate.py

Added tests for ThermostatRunningState values 0 and 128 (map to OFF)

virtualbitzz committed:
Update climate.py

Added ThermostatRunningState value for 0 = Off

virtualbitzz committed:
Update climate.py

Undid previous change

virtualbitzz committed:
Update tests/components/matter/test_climate.py

Remove test for value 0

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

MartinHjelmare reviewed (approved):
Looks good!
